### PR TITLE
CT-2266 Close offender SAR case

### PR DESCRIPTION
**Pull request by Polsonby**
## Description
Use existing pattern for closing a case instead of "Mark as closed":

## Self-review checklist

* [x] (1) Quick stakeholder demo done OR
* [ ] (2) ...bug with before and after screenshots
* [x] (3) Tests passing
* [x] (4) Branch ready to be merged (not work in progress)
* [x] (5) No superfluous changes in diff
* [x] (6) No TODO's without new ticket numbers
 
### Screenshots
image

image

  
### Related JIRA tickets
https://dsdmoj.atlassian.net/browse/CT-2266
 

### Manual testing instructions
This builds on the previous work to provide manual buttons for stepping a case through the workflow states all the way to closed. To test the new feature:

1. Create a new Offender SAR case as Brian Rix

2. Take it through to "ready to dispatch" 

3. The action button should say "Close"

4. Closing should be a case of clicking close; supply the date of dispatch and then the outcome category. 

5. Case should no longer appear in Open Cases

6. Case should now appear in Closed Cases and the case history should reflect the change 

7. The time taken should be calculated and displayed on the Case page and the outcome reason should be displayed

image

### CODE REVIEW

File: app/controllers/cases/offender_sar_controller.rb [hunk 1]
```diff
@@ -19,13 +19,6 @@ def new
       permitted_correspondence_types
       authorize case_type, :can_add_case?
 
-      @case = OffenderSARCaseForm.new(session)
-      step = params[:step].present? && params[:step] != 'new' ? params[:step] : @case.steps.first
-
-      # @todo: Why does current_step need to be set? is it something OffenderSARCaseForm
-      # can figure out itself?
-      @case.current_step = step
-
```
Comment (Polsonby) [inline app/controllers/cases/offender_sar_controller.rb]: Not related to this PR per se, but cleaned up some redundant code, partly left from merging work

**Review by mseedat-moj — CHANGES_REQUESTED** [review]
almost there minor requests

File: app/controllers/cases/offender_sar_controller.rb [hunk 2]
```diff
@@ -82,13 +74,15 @@ def cancel
...
+      authorize @case, :transition?
+
       available_actions = %w[
         mark_as_waiting_for_data
         mark_as_ready_for_vetting
         mark_as_vetting_in_progress
         mark_as_ready_to_copy
         mark_as_ready_to_dispatch
-        mark_as_closed
+        close
```
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/controllers/cases/offender_sar_controller.rb]: This work was to change to use an existing behaviour with an existing name. We don't want to change it everywhere else :)

File: app/controllers/cases/offender_sar_controller.rb [hunk 3]
```diff
@@ -5,7 +5,6 @@ class OffenderSarController < CasesController
     include GovUKDateFixes
```
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/controllers/cases/offender_sar_controller.rb]: Let's do that, but not in this PR 👍

File: app/controllers/cases/offender_sar_controller.rb [hunk 4]
```diff
@@ -5,7 +5,6 @@ class OffenderSarController < CasesController
...
 
     before_action :set_case_types, only: [:new, :create]
-    before_action :set_case, except: [:new, :create]
     before_action :set_date_of_birth, except: [:new, :create]
```
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/controllers/cases/offender_sar_controller.rb]: Separate PR IMO

File: app/views/cases/offender_sar/_closure_outcomes_form.html.slim [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  .js-info-held-status
+    = f.radio_button_fieldset :info_held_status_abbreviation, choices: CaseClosure::InfoHeldStatus.all, text_method: :name, value_method: :abbreviation
+
+  .js-outcome-group
+    = f.radio_button_fieldset :outcome_abbreviation, choices: CaseClosure::Outcome.active, text_method: :name, value_method: :abbreviation
+
+  .grid-row
+    .column-two-thirds
+      .button-holder
+        = f.submit submit_button, {class: 'button'}
```
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/views/cases/offender_sar/_closure_outcomes_form.html.slim]: Fixed, thanks

File: app/views/cases/offender_sar/_closure_outcomes_form.html.slim [hunk 6]
```diff
@@ -0,0 +1,18 @@
+- url = defined?(form_url) ? form_url : polymorphic_path(@case, action: :process_closure)
+= form_for kase, as: :"#{@correspondence_type_key}", url: url do |f|
+
+  - if @case.responded_late?
+    = render partial: 'cases/shared/team_caused_late_response', locals: {kase: @case, f: f}
+
+
```
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/views/cases/offender_sar/_closure_outcomes_form.html.slim]: fixed, thanks

File: app/controllers/cases_controller.rb [hunk 7]
```diff
@@ -10,6 +10,7 @@ class CasesController < ApplicationController
     :show,
     :destroy,
     :confirm_destroy,
+    :transition
```
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/controllers/cases_controller.rb]: Done :) 👍

File: app/controllers/cases/offender_sar_controller.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/controllers/cases/offender_sar_controller.rb]: Hah - this is funny - turned out the order of the before action statements was what was messing me up, in particular the set date of birth one - so removed as originally proposed!

File: app/controllers/cases/offender_sar_controller.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/controllers/cases/offender_sar_controller.rb]: See other comment - turned out this WAS an excellent time to do it, thanks!

**Review by Polsonby — COMMENTED** [review]
@mseedat-moj Implemented suggestions, please can you re-review?

File: app/controllers/cases/offender_sar_controller.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (Polsonby) [reply app/controllers/cases/offender_sar_controller.rb]: Turns out to be an awkward place to put a comment - I think it's fine without the comment tbh 👍

**Review by mseedat-moj — APPROVED** [review]
Top bommy

**Pull request merged**